share/doc/wake/basic_wake_tutorial.md

```diff
@@ -0,0 +1,809 @@
+
+* Function adda takes an argument of type string and concatenates it with b and gives the output. `{ }` is the string interpolation operator
+```
+def adda s = "{s}b"
+```
+
+* Function times4 takes an argument x and applies the function double twice on x
+```
+def times4 x = double (double x)
+```
+
+* Adds 2 integer values and gives out the result
+```
+def sumadd x y = x+y 
+```
+
+* power 2 3 = 8
+```
+def power x n = x^n
+```
+
+#### lambda operator (\)
+The lambda operator can be used to define paramaters of a function. 
+
+* Function aveI(averageOfInteger) takes 2 arguments x and y and adds two values and divide the result by 2. This function shows the other was of defining the parameters using lambda operator. The traditional way of defining the parameter is `def aveI x y = (x+y)/2`. Either of them are logically same.
+```
+def aveI = \x\y (x+y)/2
+```
+
+* Function aveR(averageOfReal) takes 2 arguments of type double and divides the result by 2
+```
+def aveR = \x\y (x +. y) /. 2.0
+```
+
+* Function aveM(Modulus) adds 2 values and provides the remainder 
+```
+def aveM = \x\y (x+y) % 2
+```
+
+### Wake Built-In Functions
+Few of the commonly used built-in functions in this tutorial is given below
+#### Operation on String
+* `explode` breaks a string into a list of string character.
+* `strlen` returns the length of the given string.
+* `cat` concatenates the elements in list of string to a string. 
+#### Operation on List
+* `take` takes first `n` elements in the list.
+* `drop` drops first `n` elements in the list.
+* `reverse` reverses the elements in the list.
+* `head` returns the first element in the list.
+* `tail` returns the list by removing the `head`.
+* `prepend` places the given data at the begining(head) of the given list.
+* `append` appends the given data at the end of the given list.
+* `++` concatenates any number of list into a single list.
+* `len` returns the length of the given list.
+
+* Function duplicate concantenates or copies a given string twice
+```
+def duplicate s  ="{s}{s}"
+```
+
+* integerToUnicode is a `wake built-in function` which takes the ASCII input and provides its equivalent character
+```
+wake -x 'integerToUnicode 99'
+```
+
+* unicodeToInteger is a `wake built-in function` which gives the ASCII value of the given character.
+```
+wake -x 'unicodeToInteger "a"'
+```
+
+* strlen is a `wake built-in function` which returns the length of the given string.
+```
+wake -x 'strlen "abcde"'
+```
+
+### List
+A List is a sequence of items of same type. Empty list returns `Nil`. `head` returns the first element in the list. `tail` returns the rest of the list without the `head`.
+
+* Example creates a list of given elements
+```
+def listcreate xx = xx,xx,xx,Nil
+```
+ 
+* Example shows how to prepend the list of integers with the given value
+```
+def d = 1,2,3,Nil
+def listprepend data1 = prepend data1 d
+```
+
+* concantenates 2 lists using ++ operator
+```
+wake -x '(1,2,3,Nil) ++ (4,5,6,Nil)'
```

Comment:
whats the purpose of the parentheses? Do I need them? we haven't really explained the role of parens in wake

Reply:
Parenthesis in wake is used to enforce a specific order of evaluation(precedence of operators) in expressions.

Reply:
Can you say that? what if I wrote:

wake -x '1,2,3,Nil ++ 4,5,6,Nil'

would it still work?

Reply:
No this will throw an error